### PR TITLE
Finalize the initial project structure

### PR DESCRIPTION
**Pull request by bswck**

### CODE REVIEW

File: pyjelly/options.py [hunk 1]
```diff
@@ -41,54 +44,60 @@ def register_mimetypes(extension: str = ".jelly") -> None:
...
+            logical_type_name = jelly.LogicalStreamType.Name(self.logical_type)
+            return f"StreamTypes({physical_type_name}, {logical_type_name})"
+        return f"StreamTypes({self.physical_type}, {self.logical_type})"
 
-class ConsumerStreamOptions(StreamOptions):
-    physical_type: jelly.PhysicalStreamType
-    logical_type: jelly.LogicalStreamType
+    def __post_init__(self) -> None:
+        if self.physical_type == jelly.PHYSICAL_STREAM_TYPE_UNSPECIFIED:
+            msg = "physical type must be specified"
```
Comment (Karolina-Bogacka) [inline pyjelly/options.py:83]: RIOT writer for Jelly in Jena supports physical type inference (based on first row read after the options the writer sees it as either a flat triple or flat quad stream). Do we want to support a similar functionality? If so, maybe it would make more sense to have an additional category of Exceptions (like, `PyJellyUnsupportedError`)

File: pyjelly/parse/decode.py [hunk 2]
```diff
@@ -1,32 +1,68 @@
...
+from pyjelly.options import LookupPreset, StreamOptions, StreamTypes
+from pyjelly.parse.lookup import LookupDecoder
+
+
+def options_from_frame(
+    frame: jelly.RdfStreamFrame,
+    *,
+    delimited: bool,
+) -> StreamOptions:
+    row = frame.rows[0]
```
Comment (Karolina-Bogacka) [inline pyjelly/parse/decode.py:18]: Is there ever a possibility where rows are empty? Do we want to handle that in some way?

File: pyjelly/parse/ioutils.py [hunk 3]
```diff
@@ -52,26 +54,33 @@ def delimited_jelly_hint(header: bytes) -> bool:
...
+            raise JellyConformanceError(msg)
+
         options = options_from_frame(first_frame, delimited=True)
-        return options, chained_frame_iterator(first_frame, input_stream)
+        return options, chain((first_frame,), frames)
+
+    frame = parse(jelly.RdfStreamFrame, inp.read())
+
+    if not frame.rows:
+        msg = "The stream is corrupted (only contains an empty frame)"
```
Comment (Karolina-Bogacka) [inline pyjelly/parse/ioutils.py:82]: I guess we want to catch the empty frames here?

File: pyproject.toml [hunk 4]
```diff
@@ -47,7 +47,7 @@ extras = ["pyjelly[rdflib]"] # workaround for astral-sh/uv#8607
 types = ["types-protobuf>=5.29.1.20250315"]
 
 [tool.uv]
-required-version = "==0.6.17"
+required-version = "==0.7.3"
```
Comment (Karolina-Bogacka) [inline pyproject.toml:50]: idk if I like the uv being bumped in this same GIGA pr, but not a big deal

**Review by Karolina-Bogacka — CHANGES_REQUESTED** [review]
The structure overall does look more purposeful than what we've had here before, so kudos for that. Personally I would prefer a bit more comments/ docstrings in the internals, like `rdflib/parse.py`, but it may be biased by the PR being so big it is hard to exactly find the context in which we want to use X.

File: pyproject.toml (hunk 4 above)
Comment (bswck) [reply pyproject.toml:50]: Yes, I should have removed that and moved it to a separate one, forgot about that.

File: pyjelly/parse/ioutils.py (hunk 3 above)
Comment (bswck) [reply pyjelly/parse/ioutils.py:82]: There can be only one empty frame, because it's a non-delimited stream.

File: pyjelly/parse/decode.py (hunk 2 above)
Comment (bswck) [reply pyjelly/parse/decode.py:18]: No. This accepts a frame from the iterator that filters empty-rows frames out. The contract of this function is the frame is never empty.

File: pyjelly/options.py (hunk 1 above)
Comment (bswck) [reply pyjelly/options.py:83]: It could be a subclass of `NotImplementedError`, yes.

File: pyjelly/integrations/rdflib/serialize.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            return self.encode_default_graph()
+
+        if isinstance(term, rdflib.URIRef):
+            return self.encode_iri(term)
+
+        if isinstance(term, rdflib.Literal):
+            return self.encode_literal(
+                lex=term,
+                language=term.language,
+                # Workaround RDFLib breaking LSP
```
[the remark this answers is not in the document]
Comment (bswck) [reply pyjelly/integrations/rdflib/serialize.py:30]: Sure

[the remark this answers is not in the document]
Comment (Karolina-Bogacka) [reply pyjelly/integrations/rdflib/serialize.py:30]: Ok, then please make it so

File: pyjelly/options.py (hunk 1 above)
Comment (Karolina-Bogacka) [reply pyjelly/options.py:83]: Ok, I think this is a change minor enough that we can do it now (just add the new exception to `errors.py`)

File: pyproject.toml (hunk 4 above)
Comment (Karolina-Bogacka) [reply pyproject.toml:50]: Ok, let's leave it as is, this is not a big deal

**Review by Karolina-Bogacka — APPROVED** [review]

**Comment by bswck** [comment]
After investigation, the [remaining `assert 42 == 40` errors](https://github.com/Jelly-RDF/pyjelly/actions/runs/15028164367/job/42234007071#step:6:790) in the quad tests are **false negatives**.

The reported problem is that there are always 42 input quads to serialize from [`nq-syntax-tests.nq`](https://github.com/Jelly-RDF/pyjelly/blob/4c6a3b31c91ed2dfd3c4df34f9a74d8fbcbc5e09/tests/e2e_test_cases/quads_rdf_1_1/nq-syntax-tests.nq) but only 40 are present in the parsed dataset.

The quads "missing" from the parsed dataset are:

```
<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> <http://example/g> .
<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> _:g .
```

If we stripped `http://www.w3.org/2001/XMLSchema#string` from them, we would repeat two pre-existing quads:

```diff
<http://example/s> <http://example/p> "o" <http://example/g> .
<http://example/s> <http://example/p> "o"@en <http://example/g> .
-<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> <http://example/g> .
+<http://example/s> <http://example/p> "o" <http://example/g> .
<http://example/s> <http://example/p> <http://example/o> _:g .
_:s <http://example/p> <http://example/o> _:g .
<http://example/s> <http://example/p> _:o _:g .
<http://example/s> <http://example/p> "o" _:g .
<http://example/s> <http://example/p> "o"@en _:g .
-<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> _:g .
+<http://example/s> <http://example/p> "o" _:g .
```

But in RDFLib, these two quads don't compare equal(!):

```
<http://example/s> <http://example/p> "o"
<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> 
```

which is in contrary to Jelly--we cannot assign `http://www.w3.org/2001/XMLSchema#string` an entry in the datatype lookup table[^1]--therefore no distinction between implicit and explicit `xsd:string`-typed literals is conveyed from the producer to the consumer.

Therefore, since these two quads:

```
<http://example/s> <http://example/p> "o"
<http://example/s> <http://example/p> "o"^^<http://www.w3.org/2001/XMLSchema#string> 
```

are always serialized and parsed as:

```
<http://example/s> <http://example/p> "o"
<http://example/s> <http://example/p> "o" 
```

they do compare equal, unlike before serialization.

And because RDFLib default memory store removes duplicates, these 2 equal pairs are truncated.
I checked that `dataset.add()` is called 42 times, but the resulting dataset has 40 quads.

## Possible solutions

1. Normalize the input dataset--ensure literals in the tested graph don't contain an explicit `xsd:string` annotation.

* Requires mutating the original input. Not cool.

2. Develop a context-aware, graph-aware memory store that doesn't remove duplicate entries.

* Focuses on a memory store that isn't available to other users.

3. Tweak the test to not check for cardinality, but for after-normalization equality of particular elements.

* Involves writing a custom function that accepts this edge case and avoids the problems from 1 and 2.

I'd prefer 3. Feels the most universal.

[^1]: [_If no field in the literalKind oneof is set, then the literal MUST be interpreted as a simple literal (has datatype IRI equal to http://www.w3.org/2001/XMLSchema#string)_](https://jelly-rdf.github.io/1.1.x/specification/serialization/#literals)

**Comment by bswck** [comment]
The remaining failures are not inherent to this PR and are tracked in #99.

**Comment by Ostrzyciel** [comment]
This is a bug in RDF support by rdflib. According to RDF 1.1, a literal with no dt and a literal with dt=xsd:string are exactly the same. So just make a workaround for rdflib. It's their problem.

**Review by Karolina-Bogacka — APPROVED** [review]
